sdk/resource/resource_test.go

```diff
@@ -452,6 +452,36 @@ func TestNew(t *testing.T) {
 	}
 }
 
+func TestWithHostID(t *testing.T) {
+	mockHostIDProvider()
+	t.Cleanup(restoreHostIDProvider)
+
+	ctx := context.Background()
+
+	res, err := resource.New(ctx,
+		resource.WithHostID(),
+	)
+
+	require.NoError(t, err)
+	require.EqualValues(t, map[string]string{
+		"host.id": "f2c668b579780554f70f72a063dc0864",
+	}, toMap(res))
+}
+
+func TestWithHostIDError(t *testing.T) {
+	mockHostIDProviderWithError()
+	t.Cleanup(restoreHostIDProvider)
+
+	ctx := context.Background()
+
+	res, err := resource.New(ctx,
+		resource.WithHostID(),
+	)
+
+	require.Error(t, err)
```

Comment:
If `mockHostIDProviderWithError` is updated to return [`assert.AnError`](https://pkg.go.dev/github.com/stretchr/testify/assert#pkg-variables), this can be updated to test the error is a definite value. That way when this errors for a different reason the test will not succeed when it shouldn't.

Reply:
I attempted this change. What I found was that the error message from `mockHostIDProviderWithError` as part of a different error returned by `resource.New`. See:

https://github.com/open-telemetry/opentelemetry-go/blob/264becee2751862032aad7f9c97207dac23e9009/sdk/resource/resource.go#L61

The output for my failed test looked like:

```
Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,3 +1,3 @@
        	            	 (*errors.errorString)({
        	            	- s: (string) (len=40) "assert.AnError general error for testing"
        	            	+ s: (string) (len=63) "detecting resources: [assert.AnError general error for testing]"
        	            	 })
```

Let me know if there is a way to use `assert.AnError` in this situation.

Reply:
Oh interesting, that error handling in the `New` resource does not look appropriate or correct. It should return the detect error before trying the merge, which should be its own dedicated error.

I'll put in a PR to address that, but don't want to block this on that fix. Can you add a `TODO` here describing what you did in the change to error testing and we can follow up on this afterwards?

Reply:
https://github.com/open-telemetry/opentelemetry-go/pull/3844

Reply:
I rebased to pick up the changes from #3844 and addressed the `TODO`.